python/dgl/data/movielens.py

```diff
@@ -0,0 +1,678 @@
+            )
+            print(
+                "\tTrain rating pairs : {}".format(
+                    int(self.train_graph.num_edges() / 2)
+                )
+            )
+            print(
+                "\tValid rating pairs : {}".format(
+                    int(self.valid_graph.num_edges() / 2)
+                )
+            )
+            print(
+                "\tTest rating pairs  : {}".format(
+                    int(self.test_graph.num_edges() / 2)
+                )
+            )
+
+    def __getitem__(self, idx):
+        assert (
+            idx == 0
+        ), "This dataset has only one set of training, validation and testing graph"
+        if self._transform is None:
+            return self.train_graph, self.valid_graph, self.test_graph
+        else:
+            return (
+                self._transform(self.train_graph),
+                self._transform(self.valid_graph),
+                self._transform(self.test_graph),
+            )
+
+    def __len__(self):
+        return 1
+
+    @property
+    def raw_path(self):
+        if self.name == "ml-10m":
+            return os.path.join(self.raw_dir, "ml-10M100K")
+        return os.path.join(self.raw_dir, self.name)
+
+    @property
+    def graph_path(self):
+        return os.path.join(self.raw_path, self.name + ".bin")
+
+    @property
+    def feat_path(self):
+        return os.path.join(self.raw_path, self.name + "_feat.pkl")
+
+    @property
+    def info_path(self):
+        return os.path.join(self.raw_path, self.name + ".pkl")
+
+    def _process_user_fea(self, user_data):
+        if self.name == "ml-100k" or self.name == "ml-1m":
+            ages = user_data["age"].values.astype(np.float32)
+            gender = (user_data["gender"] == "F").values.astype(np.float32)
+            all_occupations = set(user_data["occupation"])
+            occupation_map = {ele: i for i, ele in enumerate(all_occupations)}
+            occupation_one_hot = np.zeros(
+                shape=(user_data.shape[0], len(all_occupations)),
+                dtype=np.float32,
+            )
+            occupation_one_hot[
+                np.arange(user_data.shape[0]),
+                np.array(
+                    [occupation_map[ele] for ele in user_data["occupation"]]
+                ),
+            ] = 1
+            user_features = np.concatenate(
+                [
+                    ages.reshape((user_data.shape[0], 1)) / 50.0,
+                    gender.reshape((user_data.shape[0], 1)),
+                    occupation_one_hot,
+                ],
+                axis=1,
+            )
+        elif self.name == "ml-10m":
+            user_features = np.zeros(
+                shape=(user_data.shape[0], 1), dtype=np.float32
+            )
+        else:
+            raise NotImplementedError
+        return user_features
+
+    def _process_movie_fea(self, movie_data):
+        import torchtext
+        from torchtext.data.utils import get_tokenizer
```

Comment:
Is it possible to pre-extract the embeddings to some files and upload all data to DGL's s3 bucket? In other words, we can use our own URLs rather than the raw URLs.

Reply:
Yes. But I think we still need the tokenizer of torchtext.

Reply:
Why is this the case? You can pre-process all features. After that there is no need for tokenizer and torchtext.

Reply:
I'm a little confused. Can we just preprocess graphs and features in advance, upload them to s3, and let API download them when users use this dataset? If true, API would not process raw data the first time users use it. Then do we need the function process() if no preprocessing is required at the user side? Also, I suppose that implementations of all datasets can follow this paradigm, right? So when do we need process()?

Reply:
First, the graphs and features can be stored in various forms. To handle different forms of data, it can still be beneficial to have a `process` function.

Second, this can be a trade-off. When we expose all pre-processing code, the implementation is less opaque and users can understand how the dataset is processed. On the other hand, this can introduce complex dependencies as in this case.

Reply:
Ok. I got your point. Here only the movie features require torchtext. I think we can preprocess movie features, upload them to s3, remove those preprocessing codes of movie features from the API, and leave other parts unchanged. What do you think?

Reply:
Let's first upload all files including the preprocessed features to s3 and then walk through the rest implementation. Ping me offline if you need help in uploading the files.

Reply:
I have updated the code logic. Movie features would be extracted from somewhere, such as s3. Other code logics are kept unchanged. In the latest commit, I assume that movie features are stored in a certain source folder, and will be copied to the folder storing preprocessed stuff. The latest commit remove dependencies on some third-party frameworks, such as torchtext.